app/routes/user/endorsements.server.ts

```diff
@@ -0,0 +1,293 @@
+  endorserEmail: string
+  endorserSub: string
+  requestorEmail: string
+  requestorSub: string
+  created: number
+  status: EndorsementState
+}
+
+export type EndorsementState = 'approved' | 'pending' | 'rejected' | 'reported'
+
+const cognitoIdentityProviderClient = new CognitoIdentityProviderClient({})
+
+export class EndorsementsServer {
+  #sub: string
+  #currentUserEmail: string
+  #currentUserGroups: string[]
+
+  private constructor(sub: string, requestorEmail: string, groups: string[]) {
+    this.#sub = sub
+    this.#currentUserEmail = requestorEmail
+    this.#currentUserGroups = groups
+  }
+
+  static async create(request: Request) {
+    const user = await getUser(request)
+    if (!user)
+      throw new Response(null, {
+        statusText: 'not signed in',
+        status: 403,
+      })
+    return new this(user.sub, user.email, user.groups)
+  }
+
+  /**
+   * Request a pending endorsement.
+   *
+   * Throws an HTTP error if:
+   *   - The current user already belongs to the GCN Circulars submitters group
+   *   - User `endorserSub` does not exist
+   *   - User `endorserSub` is not in the GCN Circulars submitters group
+   *   - There is an existing endorsement with the same requester and submitter sub but the state is not `pending`
+   */
+  async createEndorsementRequest(endorserSub: string) {
+    if (endorserSub === this.#sub)
+      throw new Response(null, {
+        statusText:
+          'Users cannot request themselves as the endorser of their own request',
+        status: 400,
+      })
+
+    const escapedEndorserString = endorserSub.replaceAll('"', '\\"')
+    const user = (
+      await cognitoIdentityProviderClient.send(
+        new ListUsersCommand({
+          UserPoolId: process.env.COGNITO_USER_POOL_ID,
+          Filter: `sub = "${escapedEndorserString}"`,
+        })
+      )
+    )?.Users?.[0]
+    if (!user)
+      throw new Response(null, {
+        status: 400,
+        statusText: 'Requested user does not exist',
+      })
+
+    const getGroupsCommand = new AdminListGroupsForUserCommand({
+      Username: user.Username,
+      UserPoolId: process.env.COGNITO_USER_POOL_ID,
+    })
+    const groupsResponse = await cognitoIdentityProviderClient.send(
+      getGroupsCommand
+    )
+
+    if (
+      !groupsResponse.Groups ||
+      !groupsResponse.Groups.find(
+        ({ GroupName }) => GroupName === 'gcn.nasa.gov/circular-submitter'
+      )
+    )
+      throw new Response(null, {
+        statusText: 'User is not in the submitters group',
+        status: 400,
+      })
+
+    const getEndorserDataCommand = new ListUsersCommand({
+      UserPoolId: process.env.COGNITO_USER_POOL_ID,
+      Filter: `sub = "${escapedEndorserString}"`,
+    })
+
+    const response = await cognitoIdentityProviderClient.send(
+      getEndorserDataCommand
+    )
+
+    if (!response.Users)
+      throw new Response(null, {
+        statusText: 'Requested user does not exist',
+        status: 400,
+      })
```

Comment:
But you already retrieved the user above.

Reply:
duplicated code from moving stuff around. It will be removed